cmd/gateway-main.go

```diff
@@ -83,11 +91,23 @@ func mustGetGatewayCredsFromEnv() (accessKey, secretKey string) {
 //
 // - Azure Blob Storage.
 // - Add your favorite backend here.
-func newGatewayLayer(backendType, accessKey, secretKey string) (GatewayLayer, error) {
-	if gatewayBackend(backendType) != azureBackend {
-		return nil, fmt.Errorf("Unrecognized backend type %s", backendType)
+func newGatewayLayer(backendType, endpoint, accessKey, secretKey string) (GatewayLayer, error) {
+	switch gatewayBackend(backendType) {
+	case azureBackend:
+		return newAzureLayer(accessKey, secretKey)
+	case s3Backend:
+		useHTTPS := true
+		if u, err := url.Parse(endpoint); err != nil {
+			return nil, err
+		} else if u.Scheme == "http" || u.Scheme == "https" {
+			useHTTPS = (u.Scheme == "https")
+			endpoint = u.Host
+		}
+
+		return newS3Gateway(endpoint, useHTTPS, accessKey, secretKey)
```

Comment:
Current code allows urls without scheme as well. Eg. both http://play.minio.io:9000 and play.minio.io:9000 are supported now.